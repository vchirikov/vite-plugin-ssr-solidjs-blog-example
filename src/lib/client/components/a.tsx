import { type JSX, createMemo, mergeProps, splitProps } from 'solid-js';

import { usePageContext } from '#client/hooks';

export interface AnchorProps extends JSX.AnchorHTMLAttributes<HTMLAnchorElement> {
  href: string;
  /** {@link https://vite-plugin-ssr.com/clientRouting#link-prefetching keep-scroll-position} */
  keepScrollPosition?: boolean;
  /** {@link https://vite-plugin-ssr.com/clientRouting#link-prefetching data-prefetch="true"} */
  prefetch?: boolean;
  /**
   * Flag to skip client routing via
   * {@link https://vite-plugin-ssr.com/clientRouting#usage-options rel="external"}
   * it also will be skiped if you use links like `schema://url`
   */
  skipRouting?: boolean;
}

export function A(props: AnchorProps) {
  const pageContext = usePageContext();
  const [our, rest] = splitProps(props, ['href', 'keepScrollPosition', 'prefetch', 'skipRouting', 'rel']);
  const rel = !!our.rel == null
  const isActive = createMemo(() => {
    pageContext.urlPathname;
    return false;
  });

  return (
    <a
      {...rest}
      //href={href() || props.href}
      classList={{
        ...(props.class && { [props.class]: true }),
        ['active']: isActive(),
        ...props.classList
      }}
      aria-current={isActive() ? 'page' : undefined}
    />
  );
}