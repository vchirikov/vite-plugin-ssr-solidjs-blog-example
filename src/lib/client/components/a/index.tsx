import { createMemo, type JSX, splitProps } from 'solid-js';

import { useConfiguration, usePageContext } from '#client/hooks';
import type { Locales } from '#shared/i18n/i18n-types';

export interface AnchorProps extends JSX.AnchorHTMLAttributes<HTMLAnchorElement> {
  href: string;
  /** undefined == use page context locale, null = without locale, or the locale param */
  locale?: Locales | null;
  /** {@link https://vite-plugin-ssr.com/clientRouting#link-prefetching keep-scroll-position} */
  keepScrollPosition?: boolean;
  /** {@link https://vite-plugin-ssr.com/clientRouting#link-prefetching data-prefetch="true"} */
  prefetch?: boolean;
  /**
   * Flag to skip client routing via
   * {@link https://vite-plugin-ssr.com/clientRouting#usage-options rel="external"}
   * it also will be skipped if you use links which are not started from `/`
   */
  skipRouting?: boolean;
}

export function A(props: AnchorProps) {
  const pageContext = usePageContext();
  const [our, rest] = splitProps(props, ['href', 'locale', 'keepScrollPosition', 'prefetch', 'skipRouting', 'rel', 'class', 'classList']);
  const isActive = createMemo(() => pageContext.urlPathname === our.href);
  const rel = createMemo(() => {
    const arr: string[] = our.rel?.split(',') ?? [];
    if (our.skipRouting) {
      arr.push('external');
    }
    if (!our.href.startsWith('/')) {
      arr.push('nofollow', 'external');
    }
    if (our.prefetch) {
      arr.push('prefetch');
    }
    return arr.length > 0 ? arr.join(' ') : undefined;
  });
  const href = createMemo(() => {
    const href = our.href;
    // if href is an external link, we do nothing
    if (href.indexOf(':') > 0) {
      return our.href;
    }
    let locale: string;
    if (our.locale === undefined) {
      locale = `/${pageContext.locale}/`;
    }
    else if (our.locale === null) {
      locale = '/';
    }
    else {
      locale = `/${our.locale}/`;
    }
    if (href.startsWith('/')) {
      const cfg = useConfiguration();
      return `${cfg.basePath}${locale}${href.slice(1)}`;
    }
    return `${pageContext.urlParsed.pathnameOriginal}${locale}${href}`;
  });

  return (
    <a
      {...rest}
      href={href()}
      rel={rel()}
      classList={{
        ...(our.class && { [our.class]: true }),
        ['active']: isActive(),
        ...our.classList
      }}
      data-prefetch={our.prefetch}
      keep-scroll-position={our.keepScrollPosition}
      aria-current={isActive() ? 'page' : undefined} />
  );
}