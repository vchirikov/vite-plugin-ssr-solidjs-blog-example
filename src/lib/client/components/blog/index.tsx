export { Breadcrumbs } from './breadcrumbs.internal';