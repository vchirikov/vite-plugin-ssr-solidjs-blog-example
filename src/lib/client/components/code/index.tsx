export { Code } from './code.internal';
export { CodeInput } from './code-input.internal';