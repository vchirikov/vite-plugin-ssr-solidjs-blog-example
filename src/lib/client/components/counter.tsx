import { createSignal } from 'solid-js';

export function Counter() {
  const [count, setCount] = createSignal(0);
  return <button data-testid='counter' onClick={() => setCount((prev) => prev + 1)}>Counter {count()}</button>;
}