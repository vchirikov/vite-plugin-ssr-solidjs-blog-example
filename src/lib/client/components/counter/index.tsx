import { createSignal } from 'solid-js';

export function Counter() {
  const [count, setCount] = createSignal(0);
  return <button class="btn btn-primary" data-testid='counter' onClick={() => setCount((prev) => prev + 1)}>Counter {count()}</button>;
}