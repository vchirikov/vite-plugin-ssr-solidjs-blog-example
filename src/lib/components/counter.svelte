<script lang="ts">
  let counter = 0;
</script>

<button class="btn btn-xl btn-secondary" on:click={() => counter++}>Counter is {counter}</button>
