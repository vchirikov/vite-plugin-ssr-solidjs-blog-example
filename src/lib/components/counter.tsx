import { createSignal } from 'solid-js';

export function Counter() {
  const [count, setCount] = createSignal(0);
  return <button onClick={() => setCount((prev) => prev + 1)}>Counter {count()}</button>;
}