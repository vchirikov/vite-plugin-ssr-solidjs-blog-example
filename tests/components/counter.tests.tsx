import { getByTestId, render } from '@solidjs/testing-library';
import { describe, expect, it } from 'vitest';

import { Counter } from '#client/components/counter';

describe('Counter', () => {
  // it('increments value', async () => {
  //   const { queryByRole, unmount } = render(() => <Counter />);
  //   const button = (await queryByRole('button')) as HTMLButtonElement;
  //   // expect(button).toBeInTheDocument();
  //   // expect(button).toHaveTextContent(/Clicks: 0/);
  //   // fireEvent.click(button);
  //   // expect(button).toHaveTextContent(/Clicks: 1/);
  //   unmount();
  // });
  it('should render', () => {
    const { container, unmount } = render(() => <Counter />);
    expect(container).toMatchSnapshot();
    expect(getByTestId(container, 'counter')).toHaveTextContent('Counter');
    //expect(container).toHaveBeenCalled()
    unmount();
  });
});