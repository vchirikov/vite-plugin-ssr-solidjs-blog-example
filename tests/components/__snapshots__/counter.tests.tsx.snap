// Vitest Snapshot v1

exports[`Counter > should render 1`] = `
<div>
  <button
    data-testid="counter"
  >
    Counter 
    <!--#-->
    0
    <!--/-->
  </button>
</div>
`;
